pkg/domain/infra/abi/quadlet.go

```diff
@@ -0,0 +1,764 @@
+				installReport.QuadletErrors[toInstall] = err
+				continue
+			}
+			installReport.InstalledQuadlets[toInstall] = installedPath
+		default:
+			err := fileutils.Exists(toInstall)
+			if err != nil {
+				installReport.QuadletErrors[toInstall] = err
+				continue
+			}
+			// If toInstall is a single file, execute the original logic
+			installedPath, err := ic.installQuadlet(ctx, toInstall, "", installDir, assetFile, validateQuadletFile)
+			if err != nil {
+				installReport.QuadletErrors[toInstall] = err
+			} else {
+				installReport.InstalledQuadlets[toInstall] = installedPath
+			}
+		}
+	}
+
+	// TODO: Should we still do this if the above validation errored?
+	if options.ReloadSystemd {
+		if err := conn.ReloadContext(ctx); err != nil {
+			return &installReport, fmt.Errorf("reloading systemd: %w", err)
+		}
+	}
+
+	return &installReport, nil
+}
+
+// Extracts file extension from Content-Disposition or URL
+func getFileExtension(resp *http.Response) string {
+	// Try Content-Disposition header first
+	cd := resp.Header.Get("Content-Disposition")
+	if cd != "" {
+		re := regexp.MustCompile(`(?i)filename="?([^"]+)"?`)
+		matches := re.FindStringSubmatch(cd)
+		if len(matches) > 1 {
+			return path.Ext(matches[1])
+		}
+	}
+	// Fallback: use URL path
+	ext := path.Ext(resp.Request.URL.Path)
+	return ext
+}
+
+// Install a single Quadlet from a path on local disk to the given install directory.
+// Perform some minimal validation, but not much.
+// We can't know about a lot of problems without running the Quadlet binary, which we
+// only want to do once.
+func (ic *ContainerEngine) installQuadlet(_ context.Context, path, destName, installDir, assetFile string, isQuadletFile bool) (string, error) {
+	// First, validate that the source path exists and is a file
+	stat, err := os.Stat(path)
+	if err != nil {
+		return "", fmt.Errorf("quadlet to install %q does not exist or cannot be read: %w", path, err)
+	}
+	if stat.IsDir() {
+		return "", fmt.Errorf("quadlet to install %q is not a file", path)
+	}
+
+	finalPath := filepath.Join(installDir, filepath.Base(filepath.Clean(path)))
+	if destName != "" {
+		finalPath = filepath.Join(installDir, destName)
+	}
+
+	// Validate extension is valid
+	if isQuadletFile && !systemdquadlet.IsExtSupported(finalPath) {
+		return "", fmt.Errorf("%q is not a supported Quadlet file type", filepath.Ext(finalPath))
+	}
+
+	// Second, validate that the dest path does NOT exist.
+	// TODO: Overwrite flag?
+	if _, err := os.Stat(finalPath); err == nil {
+		return "", fmt.Errorf("a Quadlet with name %s already exists, refusing to overwrite", filepath.Base(finalPath))
+	}
+
+	file, err := os.OpenFile(finalPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0644)
+	if err != nil {
+		return "", fmt.Errorf("unable to open file %s: %v", filepath.Base(finalPath), err)
+	}
+	defer file.Close()
+
+	// Move the file in
+	srcFile, err := os.Open(path)
+	if err != nil {
+		return "", fmt.Errorf("unable to open file %s: %v", path, err)
```

Comment:
the error messages "stutter" since err alreayd includes the full path you are adding it again so the path will be there twice. Also again here is the %v wrapping errors, it must use %w.